keps/sig-scheduling/4328-node-affinity-based-eviction/README.md

```diff
@@ -0,0 +1,495 @@
+
+### User Stories (Optional)
+
+#### manage multi tenant cluster with dedicated nodes
+
+I have some nodes labeled with "userA=allow" and "userB=allow", which means that these nodes are available to user A and user B.
+As the scale grows I want to add new nodes only for user B while leaving the current nodes only for user A.
+I want to migrate all pods available to user B to new nodes, and I want to ensure HA of my services during the migration.
+
+Without `node-affinity-eviction`, I have to remove "userB=allow" label of node, and delete the pods manually. Also I can't use taints because they don't respect PDBs.
+With `node-affinity-eviction`, I can simply delete the "userB=allow" label from the existing nodes 
+to re-schedule all pods of user B to these new nodes.
+
+### Notes/Constraints/Caveats (Optional)
+
+N/A
+
+### Risks and Mitigations
+
+The major concern may be that during the eviction, some pods may fail to be evicted. Users can use `pod_failed_eviction_node_affinity_total` metric to find them.  
+
+Typically, the reason for a failed eviction is due to a violation of the PDB. 
+However, it is also possible that there could be issues with Kubernetes components or a misconfiguration. 
+If the eviction of the target pod fails, users should first check if the eviction violates the pod's PDB configuration. 
+If it does violate the PDB, users should manually adjust the replica setting of the owner of the pod to ensure that the eviction no longer violates the PDB configuration. Alternatively, users can manually terminate the pod.
+Another thing users can check is whether there is incorrect configuration, such as multiple PodDisruptionBudgets referencing the same Pod.
+kube-apiserver and controller-manager logs can be also checked for failures.
+
+## Design Details
+
+Introducing a new type of NodeAffinity:
+```go
+type NodeSelector struct {
+	NodeSelectorTerms []NodeSelectorTerm
+}
+
+type NodeAffinity struct {
+  // If the affinity requirements specified by this field are not met at
+  // scheduling time, the pod will not be scheduled onto the node.
+  // If the affinity requirements specified by this field cease to be met
+  // at some point during pod execution (e.g. due to an update), the system
+  // will try to eventually evict the pod from its node.
+  // For v1alpha1, evictions are performed by calling the Eviction API.
+  // This may change in later versions.
+  // +optional
+  RequiredDuringSchedulingRequiredDuringExecution *NodeSelector
+}
+```
+
+Add a controller called `node-affinity-eviction` 
+
+The controller do the following things:
+
+- Listening to the changes of node labels
+- Iterating over all pods assigned to the node(excluding mirror pods), checks the NodeAffinity field, if `RequiredDuringSchedulingRequiredDuringExecution` exists, checks if `NodeSelector` still match the new node. 
+- If `RequiredDuringSchedulingRequiredDuringExecution` is no loger met, trying to evict the pod.
```

Comment:
Beyond alpha, do we need to account for a data race, where the node label changes before the Pod is bound to the node, because the kube-scheduler didn't have that particular update?

Kubernetes should aim to be eventually consistent here.

Reply:
Did you mean: 

1. node label changed
2. pod has finished shceduling, but not bound to node yet
3. we evict the pod?

This seems correct to me.